mantidproject/mantid
# Improves usage of symmetrise tab in Data Manipulation Interface.

### Description of work
This PR  fixes some of the issues reported on #33338, in which there is a set of errors from the algorithm, but the user does not have a warning message of incorrect selected values from the GUI. Furthermore, the algorithm warning is not descriptive enough regarding the error that is produced (Incorrect range). 

While I have fixed the above problems on the GUI on this PR; There are several parts of  the symmetrise tab that could be changed to improve the user experience when attempting to invoke the algorithm from the GUI. The changes done on this GUI are summarised in the following list: 

1. The main cause of concern for non-reproducible warnings was the non-initialization of a boolean flag in the model (`m_isPositiveReflect`) that handles the symmetrise boundaries when the algorithm is called, causing undefined behaviour. This could cause the algorithm to fail seemingly randomly, as the behaviour of the algorithm depends on correct setting of the boundaries, without the user really knowing what's going on.
2. I have removed two horizontal marker lines in the embedded plot. These lines don't have any role on the algorithm parameters and could cause confusion to users (it confused me...).
3. There is a vertical line marker to indicate the centre position (E=0). This line can be moved, but it also doesn't influence the algorithm. It's a fixed marker line. The way marker lines are constructed on the mantid plot widget for C++ does not allow to change the line-style or handling of mouse events. To this end; I have added the option to include a selection of a different initial line style (dash or solid, but more could be added) from the initialization of the marker. Additionally, there is a new function in the marker class to disconnect mouse signals, so that they don't interact with mouse events, and behave simply as fixed markers. 
4. Added some handling of data validation (you can't input anymore E range values outside of the active spectra range). 
5. Up until now, whenever you change the `ReflectType` property from positive to negative, the marker lines remain on the positive side of the spectra, while only their range is changed to reflect the maximum allowed values on the negative side. But it is still possible to move markers out of range and confusing to use. I believe this is done because the symmetrise algorithm doesn't allow for negative input values, so if using negative ranges, one has to make sure they are input as absolute values. Changing this requires slightly more care when passing data from the view to the model, but I have changed the way data is sent to the model to be able to pass the markers to the negative side of the spectra when the `ReflectType` choice is set in either side (negative or positive) 
6. Error signals were not connected to the interface. Therefore showing error messages was not possible from the GUI.
7. Finally, there is a new validation step when previewing the algorithm, hinting users as to what's wrong with their selection, if they still select an invalid range.

#### Summary of work




Fixes #33338 . 




#### Further detail of work


### To test:

1. Open `Data Manipulation->Symmetrise` tab interface. 
2. On File Selector, browse for an asymmetric spectra: `'UsageData/osiris97944_graphite002_red.nxs'`
3.  Make sure `ReflectType` property is` Positive to Negative`.
4. First spectra should open and plot. You should see a vertical solid red marker line in E=0 that can't be moved. And two vertical dash marker lines that can be moved, but not outside positive spectra range. 
5. In `Elow`, `Ehigh` property fields, try inputting wrong values: You shouldn't be able to input any value outside of positve range (`0-2.15`). You shouldn't be able to put `Elow` higher than `Ehigh` and viceversa.
6.  Click on preview button to show a preview of the algorithm. 
7.  Choose `Ehigh` to be `2.0` and `Elow` to be `1.0` and click preview. An error message should appear, asking to reduce `Elow` to maximum allowed value for that spectra (`0.93`).
8. Change `ReflectType` to `Negative to Positive`.  Marker lines should go to negative side, validation and behaviour of selection should work fine.








---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
